h/services/group_list.py

```diff
@@ -147,35 +152,25 @@ def private_groups(self, user=None):
         return [group for group in user_groups if group.type == "private"]
 
     def scoped_groups(self, authority, document_uri):
-        """
-        Return scoped groups for the URI and authority
-
-        Only open and restricted groups are "supposed" to have scope, but
-        technically this query is agnostic to the group's type—it will return
-        any group who has a scope that matches the document_uri's scope.
-
-        Note: If private groups are ever allowed to be scoped, this needs
-        attention.
+        if not document_uri:
+            return []
+        matching_scopes = self._group_scope_service.fetch_by_scope(document_uri)
+        matching_scope_groupids = [scope.group_id for scope in matching_scopes]
 
-        :param authority: Filter groups by this authority
-        :type authority: string
-        :arg document_uri: Use this URI to find groups with matching scopes
-        :type document_uri: string
-        :rtype: list of :class:`h.models.group`
-        """
-        origin = scope_util.uri_scope(document_uri)
-        if not origin:
+        if not matching_scope_groupids:
             return []
 
-        groups = (
-            self._session.query(models.GroupScope, models.Group)
-            .filter(models.Group.id == models.GroupScope.group_id)
-            .filter(models.GroupScope.origin == origin)
+        # Retrieve groups for these IDs
+        scoped_groups = (
+            self._session.query(models.Group)
+            .filter(models.Group.id.in_(matching_scope_groupids))
             .filter(models.Group.authority == authority)
+            .filter(
+                models.Group.readable_by == group.ReadableBy.world
+            )  # Only "public" groups
+            .group_by(models.Group.id)  # de-dupe
```

Comment:
iiiiinteresting, really? The reasons I thought there could be dupes is that multiple scopes can match against a single URL for a single group. I shall experiment with removing this (because it _is_ kinda gross) and seeing if my tests for dupes still pass...

Reply:
The query `self._session.query(models.Group).filter(Group.id.in_(ids))` AIUI translates into `SELECT ... FROM group WHERE id IN (:ids)`, so even if an id is repeated it will still only match one row of the result.

Reply:
For example:

```
[I] (h-py3)  ~/h/r/h > ./bin/hypothesis --dev shell

Environment:
  m, models    The `h.models` module.
  registry     Active Pyramid registry.
  request      Active request object.
  session      Active database session.

>>>
>>> session.query(models.Group).filter(models.Group.id.in_(['a', 'b']))
<sqlalchemy.orm.query.Query object at 0x1139d0208>
>>> str(session.query(models.Group).filter(models.Group.id.in_(['a', 'b'])))
'SELECT "group".created AS group_created, ... \nFROM "group" \nWHERE "group".id IN (%(id_1)s, %(id_2)s)'
>>>
```